pkg/pool-weighted/contracts/InvariantGrowthProtocolFees.sol

```diff
@@ -80,5 +90,64 @@ abstract contract InvariantGrowthProtocolFees is BaseWeightedPool, ProtocolFeeCa
+        // `_beforeJoinExit` paid protocol fees accumulated between the previous and current join or exit,
+        // while this code pays any protocol fees due on the current join or exit.
+
+        // Joins and exits are symmetrical; for simplicity, we consider a join, where the invariant and supply
+        // both increase.
+
+        // |-------------------------|-- postJoinExitInvariant
+        // |   increase from fees    |
+        // |-------------------------|-- original invariant * supply growth ratio (fee-less invariant)
+        // |                         |
+        // | increase from balances  |
+        // |-------------------------|-- preJoinExitInvariant
+        // |                         |
+        // |                         |  |------------------|-- postJoinExitSupply
+        // |                         |  |    BPT minted    |
+        // |                         |  |------------------|-- preJoinExitSupply
+        // |   original invariant    |  |  original supply |
+        // |_________________________|  |__________________|
+        //
+        // If the join is proportional, the invariant and supply will likewise increase proportionally,
+        // so the growth ratios (postJoinExit / preJoinExit) will be equal. In this case, we do not charge
+        // any protocol fees.
+        //
+        // If the join is non-proportional, the supply increase will be proportionally less than the invariant increase,
+        // since the BPT minted will be based on fewer tokens (because swap fees are not included). So the supply growth
+        // is due entirely to the balance changes, while the invariant growth also includes swap fees.
+        //
+        // To isolate the amount of increase by fees then, we multiply the original invariant by the supply growth
+        // ratio to get the "feeless invariant". The difference between the final invariant and this value is then
+        // the amount of the invariant due to fees, which we convert to a percentage by normalizing against the
+        // final (postJoinExit) invariant.
+        //
+        // Compute the portion of the invariant increase due to fees
+        uint256 supplyGrowthRatio = postJoinExitSupply.divDown(preJoinExitSupply);
+        uint256 feelessInvariant = preJoinExitInvariant.mulDown(supplyGrowthRatio);
+
+        uint256 invariantDeltaFromFees = postJoinExitInvariant - feelessInvariant;
+
+        // To convert to a percentage of pool ownership, multiply by the rate,
+        // then normalize against the final invariant
+        uint256 protocolOwnershipPercentage = invariantDeltaFromFees.divDown(postJoinExitInvariant).mulDown(
+            protocolSwapFeePercentage
+        );
+
+        if (protocolOwnershipPercentage > 0) {
+            uint256 protocolFeeAmount = _calculateAdjustedProtocolFeeAmount(
+                postJoinExitSupply,
+                protocolOwnershipPercentage
+            );
```

Comment:
Shall we abstract this away? There's no difference with the ComposableStablePool code afaict so we could change this into:

```
uint256 protocolFeeAmount = _getJoinExitProtocolSwapFee(
    preJoinExitInvariant,
    postJoinExitInvariant,
    preJoinExitSupply,
    postJoinExitSupply
)
```

Reply:
I'll try; I already moved up part of the calculation into the base protocol fee contract